libopenage/renderer/opengl/renderer.cpp

```diff
@@ -79,7 +81,21 @@ void GlRenderer::render(RenderPass const& pass) {
 	glClearColor(0.0f, 0.0f, 0.0f, 0.0f);
 	glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
 
-	for (auto obj : pass.renderables) {
+	std::vector<Renderable> renderables = pass.renderables;
+	std::sort(renderables.begin(), renderables.end(), [](const Renderable& a, const Renderable& b) {
+		GLuint shader_a = ((GlUniformInput*) a.unif_in)->program->get_handle();
+		GLuint shader_b = ((GlUniformInput*) b.unif_in)->program->get_handle();
+
+		auto geom_a = dynamic_cast<GlGeometry const*>(a.geometry);
+		auto geom_b = dynamic_cast<GlGeometry const*>(b.geometry);
+		GLuint vao_a = (geom_a)? geom_a->get_vao_handle() : 0;
+		GLuint vao_b = (geom_b)? geom_b->get_vao_handle() : 0;
+		if (shader_a == shader_b)
+			return vao_a < vao_b;
```

Comment:
If I remember my own code correctly (not necessarily true, so please double-check if you can), this will not work, for two reasons:
1. In general, not every `GlGeometry` has an underlying mesh. `geometry_t::bufferless_quad` ones are generated at runtime, as will be other ones (ideally we would have more basic polygons). So the call to `get_vao_handle()`.
2. As it stands now, the constructor `GlGeometry` will make a new vertex array (VAO) every time you call it, even with the same `MeshData`, and there is no way to deduplicate them. So no VAO is shared and the ordering does not help.

I think the first step here would be to introduce deduplication (either automatic via something like hashing) at the `GlGeometry` level.

On the other hand, the ordering by shaders should help. So a PR doing just `shader_a < shader_b` would be useful.

Reply:
GitHub is refusing to let me edit that. Should say "the call to `get_vao_handle()` can fail".

Reply:
And "either automatic" -> "hopefully automatic".

Reply:
> introduce deduplication (either automatic via something like hashing) at the GlGeometry level.

does that means that `GlGeometry` would have to keep track of all the `GlMesh`es it creates?

Also, after implementing that, wouldn't the ordering by VAOs start to help automatically?

Reply:
> does that mean that GlGeometry would have to keep track of all the GlMeshes it creates?

Hm no, that doesn't sound right - my bad. It should be the responsibility of `GlRenderer` to do the deduplication, perhaps in `add_mesh_geometry`. This will likely require some plumbing in `GlGeometry` - for example replacing the owned `GlVertexArray` with a reference (`shared_ptr` i guess) to one - but the main functionality should be part of `GlRenderer`, since `GlGeometry` is only meant to manage a single object. It would have to detect 'MeshData' reuse somehow, but then the detection mechanism should be fast enough not to outweight the performance gains (if any.. afaik switching VAOs often has little to no cost, unlike switching shaders).

> Also, after implementing that, wouldn't the ordering by VAOs start to help automatically?

Yeah, sounds like it should.

Reply:
One more thing. For a long list of renderables, it would be better to sort them in-place within the `RenderPass` object rather than on every frame. Two alternatives: either change the `pass` parameter type to `RenderPass&` and update the documentation to note that the object might change, or add a `Renderer::optimize(RenderPass& pass)` function which does this separately from `render`.

Reply:
> It should be the responsibility of GlRenderer to do the deduplication, perhaps in add_mesh_geometry.

My idea is that we:
* add a hash function to `MeshData`
* link the index buffer with the VAO (it's not linked now)
* and in the place where we optimize the `Renderpass` replace all the renderables that have the same `MeshData` hash and shader handle with the same `Renderable`

Reply:
At a first glance, this looks like it should work. Note though that in principle if the hash matches, we should still check the `MeshData` for exact equality in case of collision. However, that would be quite slow, so if the hash is long enough (64/128-bit?), I think we can get away with assuming no collisions.

Reply:
This looks like it should work. However, for now I don't think it's necessary since for one, we can't actually reorder by VAO (again, sorry about the confusion) and two, deduplication can easily be done by users of the renderer API since multiple renderables can share a `Geometry*`.